sys/console/full/src/console.c

```diff
@@ -248,6 +248,186 @@ del_char(char *pos, uint16_t end)
+struct console_hist {
+    struct console_input buffer[MYNEWT_VAL(CONSOLE_HIST_SIZE)];
+    uint8_t head;
+    uint8_t tail;
+    uint8_t size;
+    uint8_t curr;
+} console_hist;
+
+static void
+console_hist_init(void)
+{
+    struct console_hist *sh = &console_hist;
+
+    sh->head = 0;
+    sh->tail = 0;
+    sh->curr = 0;
+    sh->size = MYNEWT_VAL(CONSOLE_HIST_SIZE);
+}
+
+static size_t
+trim_whitespace(const char *str, size_t len, char *out)
+{
+    const char *end;
+    size_t out_size;
+
+    if (len == 0) {
+        return 0;
+    }
+
+    /* Trim leading space */
+    while (isspace((unsigned char)*str)) {
+        str++;
+    }
+
+    if (*str == 0) { /* All spaces? */
+        *out = 0;
+        return 0;
+    }
+
+    /* Trim trailing space */
+    end = str + strlen(str) - 1;
+    while (end > str && isspace((unsigned char)*end)) {
+        end--;
+    }
+
+    end++;
+
+    /* Set output size to minimum of trimmed string length and buffer size minus 1 */
+    out_size = (end - str) < len - 1 ? (end - str) : len - 1;
+
+    /* Copy trimmed string and add null terminator */
+    memcpy(out, str, out_size);
+    out[out_size] = 0;
+
+    return out_size;
+}
+
+static uint8_t
+ring_buf_next(uint8_t i, uint8_t size)
+{
+    return (uint8_t) ((i + 1) % size);
+}
+
+static uint8_t
+ring_buf_prev(uint8_t i, uint8_t size)
+{
+    return (uint8_t) ((i - 1) % size);
+}
+
+static bool
+console_hist_full(void)
+{
+    struct console_hist *sh = &console_hist;
+
+    return sh->head == sh->tail;
+}
+
+static void
+console_hist_next(void)
+{
+    struct console_hist *sh = &console_hist;
+
+    sh->head = (uint8_t) ring_buf_next(sh->head, sh->size);
+
+    /* buffer full, start overwriting old history */
+    if (console_hist_full()) {
+        sh->tail = (uint8_t) ring_buf_next(sh->tail, sh->size);
+    }
+}
+
+static bool
+console_hist_find(char *line)
+{
+    struct console_hist *sh = &console_hist;
+    uint8_t curr;
+
+    curr = sh->tail;
+    while (curr != sh->head) {
+        if (strcmp(sh->buffer[curr].line, line) == 0) {
+            return true;
+        }
+
+        curr = ring_buf_next(curr, sh->size);
+    }
+
+    return false;
+}
+
+static void
+console_hist_add(char *line)
+{
+    struct console_hist *sh = &console_hist;
+    char buf[MYNEWT_VAL(CONSOLE_MAX_INPUT_LEN)];
+    size_t len;
+
+    /* Reset current pointer */
+    sh->curr = sh->head;
+
+    len = trim_whitespace(line, sizeof(buf), buf);
+    if (!len) {
+        return;
+    }
+
+    if (console_hist_find(buf)) {
+        return;
+    }
+
+    strcpy(sh->buffer[sh->head].line, buf);
+    console_hist_next();
+    /* Reset current pointer */
+    sh->curr = sh->head;
+}
+
+static void
+clear_line(void)
+{
+    if (cur) {
+        cursor_backward(cur);
+    }
+    end += cur;
+    cur = 0;
+    while (cur < end) {
+        console_out(' ');
+        ++cur;
+    }
+    if (cur) {
+        cursor_backward(cur);
+    }
+    end = 0;
+    cur = 0;
```

Comment:
I think this can be substituted by simple escape sequence conbination of `ESC [G` (move to 1st column) + `ESC [K` (clear from cursor to eol)